### PR TITLE
BCE - 13592 upload packages files after deploy

### PR DESCRIPTION
**Pull request by mayshavit**
* Upload changes (package.json, yarn.lock and customHash files) only after a successful deployment for making retries more reliable.
* In artifactory option - install packages only if not exist in the artifactory
* Fix export name for more than one stack to use the artifactory in the same account
* Improve logs

### CODE REVIEW

**Review by SteveVaknin — APPROVED** [review]

File: src/config/ServerlessLayersConfig.js [hunk 1]
```diff
@@ -1,6 +1,21 @@
 class ServerlessLayersConfig {
-  constructor(options) {
+  constructor(options, slsVersion) {
+    console.log(`[ LayersPlugin ]: - options are: ${JSON.stringify(options)}, sls version is ${slsVersion}`);
+
+    if (slsVersion.startsWith('3') && Object.entries(options).length > 0) {
+      const v3Options = {};
+      options.param.forEach((v3Option) => {
```
Comment (yuvalyacoby) [inline src/config/ServerlessLayersConfig.js]: what is `options.param` ? can we guarantee it's always there? you check for options entries length, but not for the existing of `param` key

File: src/config/ServerlessLayersConfig.js [hunk 2]
```diff
@@ -1,6 +1,21 @@
...
+        console.log(`option of v3 is: ${v3Option}`);
+        const v3OptionSplitArr = v3Option.split('=');
+        // @ts-ignore
+        // eslint-disable-next-line prefer-destructuring
+        v3Options[v3OptionSplitArr[0]] = v3OptionSplitArr[1];
+      });
```
Comment (yuvalyacoby) [inline src/config/ServerlessLayersConfig.js:13]: can you explain what it does please?

File: src/config/ServerlessLayersConfig.js [hunk 3]
```diff
@@ -1,6 +1,21 @@
...
+    console.log(`[ LayersPlugin ]: - options are: ${JSON.stringify(options)}, sls version is ${slsVersion}`);
+
+    if (slsVersion.startsWith('3') && Object.entries(options).length > 0) {
+      const v3Options = {};
+      options.param.forEach((v3Option) => {
+        console.log(`option of v3 is: ${v3Option}`);
+        const v3OptionSplitArr = v3Option.split('=');
+        // @ts-ignore
+        // eslint-disable-next-line prefer-destructuring
+        v3Options[v3OptionSplitArr[0]] = v3OptionSplitArr[1];
```
Comment (yuvalyacoby) [inline src/config/ServerlessLayersConfig.js:12]:
```suggestion
        const [key, value] = v3Option.split('=');
        v3Options[key] = value;
```

File: src/config/ServerlessLayersConfig.js [hunk 4]
```diff
@@ -1,6 +1,21 @@
...
+      });
+      options = v3Options;
```
Comment (yuvalyacoby) [inline src/config/ServerlessLayersConfig.js]:
```suggestion
      options = Object.entries(options).reduce((res, curr) => {
           const [key, value] = curr;
           res[key] = value;
           return res;
      }, {})
```

suggestion for cleaner code.. double check me it works...

File: src/index.js [hunk 5]
```diff
@@ -424,9 +441,12 @@ class ServerlessLayers {
 
     if (!outputs) return null;
 
-    const logicalId = this.getOutputLogicalId(this.getLayerName());
+    const layerName = this.slsLayersConfig.shouldUseLayersArtifactory ? this.slsLayersConfig.artifactoryLayerName : this.getLayerName();
+    const exportName = this.getOutputLogicalId(layerName);
 
-    const arn = (outputs.find(x => x.OutputKey === logicalId) || {}).OutputValue;
+    console.log(`[ LayersPlugin ]: going to try and find if layer name ${layerName} with the export name ${exportName} is in outputs ${JSON.stringify(outputs)}`);
```
Comment (yuvalyacoby) [inline src/index.js]: sometimes it's better to add `null, 2` to `JSON.stringify` to make it more readable: `${JSON.stringify(outputs, null, 2)})`. but if you see the logs are readable the way it is that's ok just FYI

**Review by yuvalyacoby — APPROVED** [review]
💯

File: src/config/ServerlessLayersConfig.js (hunk 1 above)
Comment (mayshavit) [reply src/config/ServerlessLayersConfig.js]: When using sls v3, the options are not written as --key value but with --param="key=value". Added a little check for this.

File: src/config/ServerlessLayersConfig.js (hunk 4 above)
Comment (mayshavit) [reply src/config/ServerlessLayersConfig.js]: Used the reduce, the --param="key=value" still needs to be parsed... And the options.param is an array I'll just reduce it

**Pull request merged**